### PR TITLE
New features and some fixes - Variadic arguments flexibility and grouping, Negative numbers 

### LINKED ISSUE #93: Negative argument values and optional variadic arguments / options.
Hi, 
First and foremost, thank you for this fantastic project! While there are others out there, I really like this one since it's small, light, simple, and very flexible.

I am having trouble with negative values:

```bash
# cmd <offset> --reduce 
$ cmd  -20 --reduce -4 # fails as it parses -20 to -2 -0 and -4 as an option name.
``` 
Usually, in some CLI programs, something like this would be evaluated correctly:
```bash
$ cmd  "-20" --reduce "-4" # same issue
$ cmd "-20" --reduce=-4 # this one partially works 
``` 
It seems that currently, `php-cli` doesn't support direct negative values unless I'm missing something (please correct me if I'm wrong). 

I modified the `Normalizer` and the `Parser` to support that if this is truly an issue let me know, and I can create a pull request.

Also, as an artifact, now a syntax as follows is easily implemented:
```bash
# cmd <name> [counted...] --variadic [values...] --another <values...>
$ cmd john [12 45 87] --variadic [ 15 20 65 ] --another 98 65 45
# Square brackets act like group boundaries for "in the middle" variadic arguments.
# Still the last variadic argument, without square brackets, functions as normal.
```
I think it's Cool... and very useful.

Let me know if that's an acceptable improvement

### PR DESCRIPTION
**Pull request by shlomohass**
This is the result of / and fixes #93 

## What has changed:
Introduced a new `Tokenizer` class to enable a more advanced syntax to be parsed (Normalizer now only handles the value filtering)
- **Negative numbers** are now supported and will be auto-detected and not treated as options.
- As discussed, I implemented a "**variadic boundaries**" syntax, which offers more flexibility and the use of multiple variadic arguments/options.
- **a bug** with options that are defined only with a long name and are variadic `->option("--names [items...]")` is now working as expected.
- minor improvements and fixes to satisfy linters and phpstan.

```bash
# Now, variadic arguments can be explicitly used : 
$ app cmd arg11 arg12 --opt [ val1 val2 ] # 👍 
$ app cmd [arg11 arg12] --opt [val1 val2] # 👍 
$ app cmd arg11 arg12 --opt val1 val2 # 👍 
$ app cmd [arg11 arg12] [arg21 arg22] --opt val1 val2 # 👍 
$ app cmd arg1 arg2 --flags [ -- -a -b -c ] --opt val1 val2 # amazingly this one is working also 👍 

#Negative numbers:
$ app scan --offset -5 # 👍 
```

## Tested on/with:
- I made a new **test file**, `AdvacedArgsTest.php` with many edge cases and crazy scenarios.
- win 10 - PHP 8.1 - bash, cmd
- ubuntu 22 - PHP 8.2
- mint - PHP 8.2
- PHPunit and real applications with user input.

## Notes:
- I did my best to keep **backward compatibility** so it should work with older implementations. That said, you should consider (in case you approve this pull) **bumping the major version** as it introduces new features and, in some rare and weird cases, it may break the current implementations - maybe someone relies on the "last variadic" error :astonished:
- I tested it in real-world applications (mine), and it worked out of the box - It will be great if you can test it against your own apps just to be sure.

I have some more ideas for future improvements 😄

### CODE REVIEW

**Comment by adhocore** [comment]
oh, just saw this and looks great. i will check indepth from PC soon. thanks for working out the PR

**Comment by shlomohass** [comment]
@adhocore Take your time and test it carefully. I really did my best to cover any weird (sensible) combination, but you can never be sure.

File: src/Helper/Polyfill.php [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+ * and still maintaining backword compatibility
+ * 
+ * @author  Shlomo Hassid <shlomohassid@gmail.com>
+ * @license MIT
+ *
+ * @link    https://github.com/adhocore/cli
+ */
+class Polyfill
+{
+    public static function str_contains($haystack, $needle)
```
Comment (adhocore) [inline src/Helper/Polyfill.php:25]: is it really needed for php8? what is the target php version for new changes?

File: src/Input/Option.php [hunk 2]
```diff
@@ -30,6 +31,10 @@ class Option extends Parameter
 
     protected string $long = '';
 
+    // We export those to be used while parsing:
```
Comment (adhocore) [inline src/Input/Option.php:34]: probably no need comment that does not help much :D

File: src/Input/Option.php [hunk 3]
```diff
@@ -41,16 +46,35 @@ protected function parse(string $raw): void
...
+        $short = '';
+        $long  = '';
+        foreach (preg_split('/[\s,\|]+/', $raw) as $part) { 
+            if (Polyfill::str_starts_with($part, self::SIGN_LONG)) {
+                $long = $part;
+            } elseif (Polyfill::str_starts_with($part, self::SIGN_SHORT)) {
+                $short = $part;
+            }
+        }
+        return [
```
Comment (adhocore) [inline src/Input/Option.php:73]: NL before return:

```
}

return [
```

File: src/Input/Parser.php [hunk 4]
```diff
@@ -54,30 +52,39 @@ abstract class Parser
...
+            if (
+                $token->isConstant() ||
+                $token->isVariadic()
+            ) {
+                $this->parseArgs($token, $this->_tokenizer);
+                continue;
             }
+
+            // Its an option parse it and its value/s:
+            if (
```
Comment (adhocore) [inline src/Input/Parser.php:81]: i think for single clause, single line is preferable:

```php
if ($token->isOption()) {
```

File: src/Input/Parser.php [hunk 5]
```diff
@@ -54,30 +52,39 @@ abstract class Parser
...
+                $token->isOption()
+            ) {
+                $this->parseOptions($token, $this->_tokenizer);
+                continue;
```
Comment (adhocore) [inline src/Input/Parser.php:85]: continue not needed as it is last statement

File: src/Input/Parser.php [hunk 6]
```diff
@@ -86,76 +93,213 @@ public function parse(array $argv): self
...
      */
-    protected function parseArgs(string $arg)
+    protected function parseArgs(Token $arg, Tokenizer $queue) : void
     {
-        if ($this->_lastVariadic) {
-            return $this->set($this->_lastVariadic, $arg, true);
+        // Handle this argument:
+        $argument = array_shift($this->_arguments);
+        
+        //No argument defined, so its an indexed arg:
```
Comment (adhocore) [inline src/Input/Parser.php:108]: for comment a space is preferable after `//`:
`// No` instead of `//No`

File: src/Input/Parser.php [hunk 7]
```diff
@@ -86,76 +93,213 @@ public function parse(array $argv): self
...
+        if (is_null($argument)) {
+
+            //Its a single constant value arg:
+            if ($arg->isConstant()) {
+
+                $this->set(null, $arg->value());
+
+            } else {
```
Comment (adhocore) [inline src/Input/Parser.php:116]: the if/else looks a bit oddly. can we refactor it? maybe a method extraction for non-constant case

File: src/Input/Parser.php [hunk 8]
```diff
@@ -86,76 +93,213 @@ public function parse(array $argv): self
...
+            }
+            return;
         }
 
-        if (!$argument = reset($this->_arguments)) {
-            return $this->set(null, $arg);
+        // Its variadic, so we need to collect all the remaining args:
+        if ($argument->variadic() && $arg->isConstant()) {
+
+            // Consume all the remaining tokens
```
Comment (adhocore) [inline src/Input/Parser.php:132]: seems the comment is short enough to fit in single line:
`// Consume all the remaining tokens if an option is found, treat it as well`

File: src/Input/Parser.php [hunk 9]
```diff
@@ -86,76 +93,213 @@ public function parse(array $argv): self
...
+                    }
+                }
+            }
+            return;
         }
 
-        if (!$argument = reset($this->_arguments)) {
-            return $this->set(null, $arg);
+        // Its variadic, so we need to collect all the remaining args:
+        if ($argument->variadic() && $arg->isConstant()) {
```
Comment (adhocore) [inline src/Input/Parser.php:130]: this block should be a different method

File: src/Input/Parser.php [hunk 10]
```diff
@@ -86,76 +93,213 @@ public function parse(array $argv): self
...
+                    
+                    throw new InvalidParameterException("Only constant parameters are allowed in variadic arguments");
+                }
+            }
+            return;
         }
 
-        $this->setValue($argument, $arg);
+        // Its variadic, and we have a variadic grouped arg:
+        if ($argument->variadic() && $arg->isVariadic()) {
```
Comment (adhocore) [inline src/Input/Parser.php:160]: should be extracted to different method

File: src/Input/Parser.php [hunk 11]
```diff
@@ -86,76 +93,213 @@ public function parse(array $argv): self
...
-    protected function parseOptions(string $arg, string $nextArg = null): bool
-    {
-        $value = substr($nextArg ?? '', 0, 1) === '-' ? null : $nextArg;
+    protected function parseOptions(Token $opt, Tokenizer $tokens) : int
+    {   
+
+        // Look ahead for next token:
+        $next = $tokens->offset(1);
+
+        // Get the option:
```
Comment (adhocore) [inline src/Input/Parser.php:201]: no need comment

File: src/Input/Parser.php [hunk 12]
```diff
@@ -86,76 +93,213 @@ public function parse(array $argv): self
...
+        $option = $this->optionFor($opt->value());
+        
+        //Consumed:
```
Comment (adhocore) [inline src/Input/Parser.php:204]: no need comment

File: src/Input/Parser.php [hunk 13]
```diff
@@ -86,76 +93,213 @@ public function parse(array $argv): self
...
+            return $consumed;
         }
 
-        $this->_lastVariadic = $option->variadic() ? $option->attributeName() : null;
+        // If option is variadic, and next is a variadic group, 
+        // then we need to collect all the nested values:
+        if ($option->variadic() && $next->isVariadic()) {
+
+            foreach ($next->nested as $token) {
+                if ($token->isConstant()) {
```
Comment (adhocore) [inline src/Input/Parser.php:249]: i see many code repetition of this nature:
```
if (token isConstant) { 
  setValue
} else {
  throw ...
}
```

--- should be extracted to some helper (`setConstantOrFail()`) that can be used like this:
```$this->setConstantOrFail($token)`

File: src/Input/Tokenizer.php [hunk 14]
```diff
@@ -0,0 +1,464 @@
...
+    /** 
+     * toString magic method.
+     * for debugging.
+     */
+    public function __toString()
+    {
+        $str = PHP_EOL;
+        foreach ($this->tokens as $token) {
+            $str .= " - ".$token . PHP_EOL;
+            if (!empty($token->nested)) {
```
Comment (adhocore) [inline src/Input/Tokenizer.php:456]: !empty check not needed unless nested can be null. but i think we can use empty array `[]` for nested as default value

File: src/Input/Tokenizer.php [hunk 15]
```diff
@@ -0,0 +1,464 @@
...
+    }
+
+    /**
+     * Get the current token if valid.
+     * 
+     * @return Token|null
+     */
+    public function validCurrent(): ?Token
+    {
+        if ($this->valid()) {
```
Comment (adhocore) [inline src/Input/Tokenizer.php:441]: return $this->valid() ? $this->current() : null;

File: src/Input/Tokenizer.php [hunk 16]
```diff
@@ -0,0 +1,464 @@
...
+     * @return bool
+     */
+    public function valid(): bool
+    {
+        return isset($this->tokens[$this->index]);
+    }
+
+    /**
+     * Rewind the pointer to the first token.
+     * For Iterator interface.
```
Comment (adhocore) [inline src/Input/Tokenizer.php:425]: `For Iterator interface.` should be removed from all methods

File: src/Input/Tokenizer.php [hunk 17]
```diff
@@ -0,0 +1,464 @@
...
+     * @param int $offset
+     * 
+     * @return Token|null
+     */
+    public function offset(int $offset): ?Token
+    {
+        if (isset($this->tokens[$this->index + $offset])) {
+            return $this->tokens[$this->index + $offset];
+        }
+        return null;
```
Comment (adhocore) [inline src/Input/Tokenizer.php:387]: oneliner with null coalesce:

`return $this->tokens[$this->index + $offset] ?? null;`

File: src/Input/Tokenizer.php [hunk 18]
```diff
@@ -0,0 +1,464 @@
...
+     */
+    private function isConstant(string $arg): bool
+    {   
+        // Early return for non-option args its a constant:
+        if (!$this->isOption($arg)) { 
+            return true;
+        }
+        // If its a single hyphen, maybe its a negative number:
+        if (
+            ($arg[0] ?? '') === '-'
```
Comment (adhocore) [inline src/Input/Tokenizer.php:133]: btw we cant assume `-N` where N is number, to be a number it can be a option as well

eg `curl` uses `-1` `-2` `-3` as option for ssl v1 to v3

File: src/Input/Tokenizer.php [hunk 19]
```diff
@@ -0,0 +1,464 @@
...
+                    case Token::TYPE_LITERAL:
+                        // Add all remaining tokens as nested of a literal token:
+                        $literal = true;
+                        break;
+
+                    case Token::TYPE_SHORT:
+                    case Token::TYPE_LONG:
+                    case Token::TYPE_CONSTANT:
+                        if ($variadic) {
+                            $this->tokens[count($this->tokens) - 1]->addNested($token);
```
Comment (adhocore) [inline src/Input/Tokenizer.php:97]: the nesting goes quite deep: `class -> func -> foreach -> foreach -> switch -> case -> if` 
probably need to optimize, specially loop inside loop

File: src/Input/Token.php [hunk 20]
```diff
@@ -0,0 +1,158 @@
...
+    public function isOption(): bool {
+        return $this->type(self::TYPE_SHORT) || $this->type(self::TYPE_LONG);
+    }
+
+    /**
+     * Get the values of the nested tokens.
+     * 
+     * @return array
+     */
+    public function nestedValues(): array {
```
Comment (adhocore) [inline src/Input/Token.php:135]:
```
public function nestedValues(): array
{
```

File: src/Input/Token.php [hunk 21]
```diff
@@ -0,0 +1,158 @@
...
+ *     <https://github.com/adhocore>
+ *
+ * Licensed under MIT license.
+ */
+
+namespace Ahc\Cli\Input;
+
+use Ahc\Cli\Input\Option;
+
+use function \array_map;
```
Comment (adhocore) [inline src/Input/Token.php:16]: `\` not needed

**Review by adhocore — COMMENTED** [review]
hello @shlomohass so too many things and big scope, i think i have left enough comments in review for now. (will check again later again with more comments lol). 
quick update we can do here besides addressing reviews; i guess, is firstly using a proper linting/php-cs-fixer. thank you

File: src/Helper/Polyfill.php (hunk 1 above)
Comment (shlomohass) [reply src/Helper/Polyfill.php:25]: No, it's not. I generally think new releases should only support 8+,  I thought you were still supporting 7.4 just realized you have a dedicated release for PHP 7.0. 
Mark it in the review I will remove it when you are done with the review 😄.

**Comment by shlomohass** [comment]
@adhocore Sure I'll start making the edits. I thought the `CodeFactor` has autofix for all the empty lines and so on... I will do it on my branch and update the pull.

File: src/Input/Parser.php (hunk 7 above)
Comment (shlomohass) [reply src/Input/Parser.php:116]: What do you mean? I'm still determining... 

I had another version which was recursive and messy.
I think the naive way of simple, expressive conditions is the simplest and easiest way (and performant) 

Can you elaborate on that?

File: src/Input/Parser.php (hunk 13 above)
Comment (shlomohass) [reply src/Input/Parser.php:249]: This is because I want to avoid recursion (as much as possible), and as a general rule I normally don't add more abstraction, especially for internal code which probably will not be used anywhere else...
That said I can add more methods if it's a design rule.

Comment (adhocore) [reply src/Input/Parser.php:249]: this loop is convoluted so if the command definitions for options don't contain any variadic option at all, we should have a way to just use older approach of parsing for performance reasons. i think variadic options are only a small percentage of non variadic options.

File: src/Input/Tokenizer.php (hunk 18 above)
Comment (shlomohass) [reply src/Input/Tokenizer.php:131]: This is pretty rare and usually confusing, I can think of a way to support numerical names for options by checking first if the current identified number is defined as an option. 
That said, it still introduces some extra complexity and weirdness : -10 -12 ...

File: src/Input/Tokenizer.php (hunk 19 above)
Comment (shlomohass) [reply src/Input/Tokenizer.php:95]: What is wrong with that? It's a lot cheaper than function/method calls...
1 scope block + 4 code blocks are not considered as deep I think.

File: src/Input/Token.php (hunk 21 above)
Comment (shlomohass) [reply src/Input/Token.php:16]: We can debate on that 😃 Anyway, I removed it.

File: src/Input/Parser.php (hunk 13 above)
Comment (shlomohass) [reply src/Input/Parser.php:230]: I'm sorry I don't really understand the point.

My reasoning is:
1. The number of tokens we will iterate through will most of the time, be meaningless, I mean it's a CLI arguments parser...
2. I wanted to reduce function calls as much as possible (for performance) and I wanted to avoid unnecessary Iterations/function calls.
3. I chose this pattern because it's simple, bulletproof, and only consumes/loops/passes values if it's absolutely needed.

In this case unless `$option->variadic()` which means it was declared as variadic and I have a specific `$next->isVariadic()` (which means it's a specific variadic group i.e. [ 1 2 3 ]) - no loop will be reached.

If this is the case all the "grouped" arguments will be consumed in one pass.

Please elaborate

**Comment by shlomohass** [comment]
image

Those Are the inputs: 
image

On:
- Ubuntu 22
- PHP 8.2.1
